src/cargo/core/profiles.rs

```diff
@@ -0,0 +1,395 @@
+        let mut result = Profile::default();
+        result.debuginfo = for_unit_profile.debuginfo;
+        result.opt_level = for_unit_profile.opt_level;
+        result
+    }
+
+    /// This returns a generic base profile. This is currently used for the
+    /// `[Finished]` line.  It is not entirely accurate, since it doesn't
+    /// select for the package that was actually built.
+    pub fn base_profile(&self, release: bool) -> Profile {
+        if release {
+            self.release.profile_for("", true, ProfileFor::Any)
+        } else {
+            self.dev.profile_for("", true, ProfileFor::Any)
+        }
+    }
+
+    /// Used to check for overrides for non-existing packages.
+    pub fn validate_packages(
+        &self,
+        shell: &mut Shell,
+        packages: &HashSet<&str>,
+    ) -> CargoResult<()> {
+        self.dev.validate_packages(shell, packages)?;
+        self.release.validate_packages(shell, packages)?;
+        self.test.validate_packages(shell, packages)?;
+        self.bench.validate_packages(shell, packages)?;
+        self.doc.validate_packages(shell, packages)?;
+        Ok(())
+    }
+}
+
+/// An object used for handling the profile override hierarchy.
+///
+/// The precedence of profiles are (first one wins):
+/// - [profile.dev.overrides.name] - A named package.
+/// - [profile.dev.overrides."*"] - This cannot apply to workspace members.
+/// - [profile.dev.build-override] - This can only apply to `build.rs` scripts
+///   and their dependencies.
+/// - [profile.dev]
+/// - Default (hard-coded) values.
+#[derive(Debug, Clone)]
+struct ProfileMaker {
+    default: Profile,
+    toml: Option<TomlProfile>,
+}
+
+impl ProfileMaker {
+    fn profile_for(&self, pkg_name: &str, is_member: bool, profile_for: ProfileFor) -> Profile {
+        let mut profile = self.default;
+        if let Some(ref toml) = self.toml {
+            merge_profile(&mut profile, toml);
+            if profile_for == ProfileFor::CustomBuild {
+                if let Some(ref build_override) = toml.build_override {
+                    merge_profile(&mut profile, build_override);
+                }
+            }
+            if let Some(ref overrides) = toml.overrides {
+                if !is_member {
+                    if let Some(star) = overrides.get("*") {
+                        merge_profile(&mut profile, star);
+                    }
+                }
+                if let Some(byname) = overrides.get(pkg_name) {
+                    merge_profile(&mut profile, byname);
+                }
+            }
+        }
+        profile
+    }
+
+    fn validate_packages(&self, shell: &mut Shell, packages: &HashSet<&str>) -> CargoResult<()> {
+        if let Some(ref toml) = self.toml {
+            if let Some(ref overrides) = toml.overrides {
+                for key in overrides.keys().filter(|k| k.as_str() != "*") {
+                    if !packages.contains(key.as_str()) {
```

Comment:
I think this is something we'll want to note on the tracking issue, I'd expect the `packages` set here to be more like a `HashSet<PackageId>` (or a `Resolve`) and the keys of `overrides` here would be a `PackageIdSpec` in theory

Reply:
Basically yeah. We've got the idea of a "package id specification" which is intended to describe a particular crate in a crate graph unbambiguously, and this would allow you to, for example, customize the compilation of `foo` from crates.io but not customize `foo` from github.com/example/foo (or something like that)